plugins/aea-ledger-cosmos/aea_ledger_cosmos/cosmos.py

```diff
@@ -1012,6 +1012,118 @@ def get_transfer_transaction(  # pylint: disable=arguments-differ
         )
         return tx
 
+    def get_packed_exec_msg(
+        self,
+        sender_address: Address,
+        contract_address: str,
+        msg: JSONLike,
+        funds: int = 0,
+    ) -> ProtoAny:
+        """
+        Create and pack MsgExecuteContract
+
+        :param sender_address: Address of sender
+        :param contract_address: Address of contract
+        :param msg: Paramaters to be passed to smart contract
+        :param funds: Funds to be sent to smart contract
+
+        :return: Packed MsgExecuteContract
+        """
+
+        if funds == 0:
+            funds_coins = []
+        else:
+            funds_coins = [Coin(denom=self.denom, amount=str(funds))]
+
+        msg_send = MsgExecuteContract(
+            sender=str(sender_address),
+            contract=contract_address,
+            msg=json.dumps(msg).encode("UTF8"),
+            funds=funds_coins,
+        )
+        send_msg_packed = ProtoAny()
+        send_msg_packed.Pack(msg_send, type_url_prefix="/")
+
+        return send_msg_packed
+
+    def get_packed_send_msg(
+        self, from_address: Address, to_address: Address, amount: int
+    ) -> ProtoAny:
+        """
+        Generate and pack MsgSend
+
+        :param from_address: Address of sender
+        :param to_address: Address of recipient
+        :param amount: amount of coins to be sent
+
+        :return: packer ProtoAny type message
+        """
+        amount_coins = [Coin(denom=self.denom, amount=str(amount))]
+
+        msg_send = MsgSend(
+            from_address=str(from_address),
+            to_address=str(to_address),
+            amount=amount_coins,
+        )
+        send_msg_packed = ProtoAny()
+        send_msg_packed.Pack(msg_send, type_url_prefix="/")
+
+        return send_msg_packed
+
+    def get_multi_transaction(
+        self,
+        from_addresses: List[str],
+        pub_keys: Optional[List[bytes]],
```

Comment:
from_addresses but pub_keys...inconsistent. It should also explain in the docstring if they have to match. If so, worth asserting that match (i.e. address derived from pub_key is the one provided)?

Reply:
In Cosmos multiple addresses can be related to one public key. So we need both. And sure, len of addresses and keys needs to be the same.

Reply:
In Cosmos one pubkey can be used to generate multiple addresses. So we need both.